### PR TITLE
CLDC-275: Form flow

### PR DESCRIPTION
**Pull request by kosiakkatrina**
- [x] Add "in progress", "completed" tags to form sections based on completed questions to each section in the task list page
image

- [ ] Add status to check answers pages:
Was done as part of the [277 ticket](https://github.com/communitiesuk/mhclg-data-collection-beta/pull/25)

- [x] Include skip to next incomplete/in progress section in tasklist
image

- [ ] Check answers page shows progress:
Was done as part of the [277 ticket](https://github.com/communitiesuk/mhclg-data-collection-beta/pull/25)

- [ ] Check answers page has working answer missing link:
Was done as part of the [277 ticket](https://github.com/communitiesuk/mhclg-data-collection-beta/pull/25)

**Anything else**
Update gitignore file
Make calculation of completed subsections dynamic (_You've completed x of x sections._)

### CODE REVIEW

File: app/helpers/tasklist_helper.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    :in_progress
+  end
+
+  def get_status_style(status_label)
+    STYLES[status_label]
+  end
+
+  def get_status_label(status)
+    STATUSES[status]
+  end
```
Comment (kosiakkatrina) [inline app/helpers/tasklist_helper.rb]: Can I access these from the module directly?

File: app/helpers/tasklist_helper.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    :not_started => "govuk-tag--grey",
+    :cannot_start_yet => "govuk-tag--grey",
+    :completed => "",
+    :in_progress => "govuk-tag--blue"
+  }
+
+  def get_subsection_status(subsection_name, case_log, questions)
+    if subsection_name == "declaration"
+      return all_questions_completed(case_log) ? :not_started : :cannot_start_yet
+    end
```
Comment (dushan-madetech) [inline app/helpers/tasklist_helper.rb:19]: You could also make this a guard clause so it matches lines 21 and 22.
 
something like:
 `return all_questions_completed(case_log) ? :not_started : :cannot_start_yet if subsection_name == "declaration"`

Up to you if you want to do it 🙂

File: app/helpers/tasklist_helper.rb (hunk 1 above)
Comment (baarkerlounger) [reply app/helpers/tasklist_helper.rb]: You should be able to do `TasklistHelper::STATUSES`?

File: app/helpers/tasklist_helper.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
+module TasklistHelper
+  STATUSES = {
+    :not_started => "Not started",
+    :cannot_start_yet =>  "Cannot start yet",
+    :completed => "Completed",
+    :in_progress => "In progress"
+  }
```
Comment (dushan-madetech) [inline app/helpers/tasklist_helper.rb]: maybe you could add `.freeze` to the end? I imagine these will be constant

File: app/helpers/tasklist_helper.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  STYLES = {
+    :not_started => "govuk-tag--grey",
+    :cannot_start_yet => "govuk-tag--grey",
+    :completed => "",
+    :in_progress => "govuk-tag--blue"
+  }
```
Comment (dushan-madetech) [inline app/helpers/tasklist_helper.rb]: same as comment for lines 2-7

**Review by dushan-madetech — APPROVED** [review]
Nothing code wise really jumped out at me as needing any immediate action so 👍 from me. I've left a couple of comments but they're all up to you if you'd like to action them 🙂

File: app/helpers/tasklist_helper.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  def get_subsection_status(subsection_name, case_log, questions)
+    if subsection_name == "declaration"
+      return all_questions_completed(case_log) ? :not_started : :cannot_start_yet
+    end
+
+    return :not_started if questions.all? {|question| case_log[question].blank?}
+    return :completed if questions.all? {|question| case_log[question].present?}
```
Comment (baarkerlounger) [inline app/helpers/tasklist_helper.rb]: Think this will get more complicated once we have conditional questions in the form but probably best to keep that as a separate PR/Ticket.

File: app/helpers/tasklist_helper.rb (hunk 2 above)
Comment (kosiakkatrina) [reply app/helpers/tasklist_helper.rb:19]: Hm, I thought about it, but it's like a double if statement, so might look a bit confusing

File: app/helpers/tasklist_helper.rb (hunk 3 above)
Comment (kosiakkatrina) [reply app/helpers/tasklist_helper.rb]: Yup! Nice one!

File: app/helpers/tasklist_helper.rb (hunk 1 above)
Comment (kosiakkatrina) [reply app/helpers/tasklist_helper.rb]: Oooh, I'll try that!

File: app/helpers/tasklist_helper.rb (hunk 5 above)
Comment (kosiakkatrina) [reply app/helpers/tasklist_helper.rb]: Yeah, this is a bit to iffy (👀) for my liking, but hopefully not to difficult to change when it comes to it

File: app/helpers/tasklist_helper.rb (hunk 2 above)
Comment (dushan-madetech) [reply app/helpers/tasklist_helper.rb:19]: You could do something like the following:
``` 
declaration_status = all_questions_completed(case_log) ? :not_started : :cannot_start_yet 
return declaration_status if subsection_name == "declaration"
```
If you think it makes the ternary statement clearer

**Review by baarkerlounger — APPROVED** [review]
